### ISSUE TITLE
Feature Request: Warnings or Errors on NumPy functions being called on CvxPy Expressions

### ISSUE THREAD

**Issue by PTNobel** [cvxpy/cvxpy#1647]
# The Issue

## Existing Errors
Consider:
```python
np.sqrt(cp.Variable())
```

On my computer right now, this code crashes the iPython interpreter and raises the following exception with an unhelpful error message:

```
AttributeError: 'Variable' object has no attribute 'sqrt'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: loop of ufunc does not support argument 0 of type Variable which has no callable sqrt method
```

## Probably wrong code

Consider the code:
```python
np.hstack([cp.Variable(3), cp.Variable(3)])
```

This creates a NumPy array with dtype=object, which is often not what users want or expect.

# My Solution
## Existing Errors

I'd like to either make this work correctly or change this to raise an error with a message along the following lines:

```
NumpyCvxpyCompatibilityError: "You're calling a NumPy function on a CvxPy expression. Please use the appropriate atom from: <link to atoms page>"
```

## Probably wrong code

We should raise a warning or an error in this situation.

# Alternatives

Raise just a warning and have the code actually work if it is an atom we also implement.

# Implementation

I'm happy to do the implementation myself; I implemented a similar thing in [PTNobel/AutoDiff](https://github.com/PTNobel/AutoDiff).

**Comment by SteveDiamond** [comment]
The error would be great. Didn't know it was possible. Can you point to how you did it in your package?

**Comment by PTNobel** [comment]
So AutoDiff actually aims to have these functions to work instead of raise an error.

So NumPy checks for the method `__array_ufunc__` or `__array_function__` and calls them before trying to run its naive implementation of this code. In AutoDiff, I have a dictionary that maps the NumPy functions and ufuncs to my implementation of these methods and then I do a dictionary lookup to effectively have dynamic dispatch. Here's the relevant code:

https://github.com/PTNobel/AutoDiff/blob/caed5e47a408e8668394d3aeabf5db7900551895/auto_diff/vecvalder.py#L60-L73

Here's an example implementation:

https://github.com/PTNobel/AutoDiff/blob/caed5e47a408e8668394d3aeabf5db7900551895/auto_diff/vecvalder_funcs_and_ufuncs.py#L124-L129

If we want to raise an Error we can basically copy these implementations and just write `raise Error`.

I know when I texted @akshayka about this, he was a little skeptical of raising errors and suggested just doing warnings.
I'm not sure he was aware that `np.sqrt` case currently errors.

**Comment by PTNobel** [comment]
Section 3.2 of this paper describes the situation in more details: https://dl.acm.org/doi/pdf/10.5555/3408207.3408219

**Comment by rileyjmurray** [comment]
Related to Parth's goal of having the functions work as opposed to raising an error: https://github.com/cvxpy/cvxpy/discussions/1645#discussion-3858894

**Comment by akshayka** [comment]
I'm okay with an error if Steven and Riley are. It just seemed a little heavy handed, but then again I wasn't aware that the sqrt function errors.

**Comment by SteveDiamond** [comment]
@PTNobel maybe you could start the PR with a warning and we'll discuss more there?

**Comment by PTNobel** [comment]
@rileyjmurray I'm a bit skeptical that plug-and-play with NumPy code is achievable. With AutoDiff, in order to be compatible with my student's code while teaching Berkeley's Numerical Simulation and Modeling class, I had to end up replacing NumPy functions like `np.zeros`, `np.array`, and replacing the class `np.ndarray` with a function that emulated its constructors behaviour.

This sucked because a lot of fancier code ends up breaking. Maybe since CVXPY is more than just me building a final project, it'll be possible to keep up with NumPy and cover a reasonable percentage of the API. I'm skeptical though.

**Comment by PTNobel** [comment]
@SteveDiamond Happy to! If I don't get it done in the next few days, I'll probably have to do it after my quals on the 23rd.

**Comment by PTNobel** [comment]
Actually, Steve do you mind if I start with an error for anything which currently errors?

Otherwise we'll end up with something like
```
WARNING: Please stop what you're doing

Error: sqrt method is missing
TypeError: ...
```

Whereas I think 
```
Error: Please stop what you're doing
```
might be better?

**Comment by PTNobel** [comment]
Also what's a good place to put tests for this?

**Comment by PTNobel** [comment]
Closed by #1648
